contracts/LinniaPermissions.sol

```diff
@@ -16,25 +16,33 @@ contract LinniaPermissions is Ownable, Pausable, Destructible {
         string dataUri;
     }
 
-    event LinniaAccessGranted(bytes32 indexed dataHash, address indexed owner,
-        address indexed viewer
+    event LinniaAccessGranted(bytes32 dataHash, address indexed owner,
+        address indexed viewer, address indexed sender
     );
-    event LinniaAccessRevoked(bytes32 indexed dataHash, address indexed owner,
-        address indexed viewer
+    event LinniaAccessRevoked(bytes32 dataHash, address indexed owner,
+        address indexed viewer, address indexed sender
     );
+    event LinniaDelegateAdded(address indexed user, address indexed delegate);
 
     LinniaHub public hub;
     // dataHash => viewer => permission mapping
     mapping(bytes32 => mapping(address => Permission)) public permissions;
+    // user => delegate => bool mapping
+    mapping(address => mapping(address => bool)) public delegates;
 
     /* Modifiers */
     modifier onlyUser() {
         require(hub.usersContract().isUser(msg.sender) == true);
         _;
     }
 
-    modifier onlyRecordOwnerOf(bytes32 dataHash) {
-        require(hub.recordsContract().recordOwnerOf(dataHash) == msg.sender);
+    modifier onlyRecordOwnerOf(bytes32 dataHash, address owner) {
+        require(hub.recordsContract().recordOwnerOf(dataHash) == owner);
+        _;
+    }
+
+    modifier onlyDelegate(address owner) {
```

Comment:
consider renaming `onlyWhenSenderIsDelegate` or something that relates to msg.sender.